src/libmeasurement_kit/report/report.cpp

```diff
@@ -0,0 +1,80 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include <cstring>
+#include <measurement_kit/report.hpp>
+#include "src/libmeasurement_kit/common/utils.hpp"
+
+namespace mk {
+namespace report {
+
+Report::Report() {
+    memset(&test_start_time, 0, sizeof (test_start_time));
+}
+
+void Report::add_reporter(Var<BaseReporter> reporter) {
+    reporters_.push_back(reporter);
+}
+
+void Report::fill_entry(Entry &entry) const {
+    entry["test_name"] = test_name;
+    entry["test_version"] = test_version;
+    entry["test_start_time"] = *mk::timestamp(&test_start_time);
+    // header["options"] = options;
+    entry["probe_ip"] = probe_ip;
+    entry["probe_asn"] = probe_asn;
+    entry["probe_cc"] = probe_cc;
+    entry["software_name"] = software_name;
+    entry["software_version"] = software_version;
+    entry["data_format_version"] = data_format_version;
+}
+
+Entry Report::get_dummy_entry() const {
+    Entry entry;
+    fill_entry(entry);
+    return entry;
+}
+
+#define FMAP mk::fmap<Var<BaseReporter>, Continuation<Error>>
+
+void Report::open(Callback<Error> callback) {
+    if (openned_) {
+        callback(ReportAlreadyOpen());
+        return;
+    }
+    openned_ = true;
```

Comment:
I can think of some issues with setting this attribute before you actually call the open report function.

If for some reason one of the two report opens fail an error message will be returned to the caller however the called doesn't have the ability to re-issue the open since it's guarded by the openned_ flag.

It's probably also not wise to put it after the call, though, since that would mean that if you call `open` while the open is in progress you will still get an error.

A possible way out of this would be to move the `openned_` and `closed_` flags into the report functions and make them a state variable instead of a boolean flag.

The states could be:

* INITIAL

* OPENNED

* IN_PROGRESS

* CLOSED

Where IN_PROGRESS is used if either an open or a close is in progress.